From 4b0e1781d6d1690fda5eab5648e7b7783295cb89 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Thu, 26 Sep 2024 16:04:09 +0200
Subject: [PATCH] serial: stm32: use pm_runtime_autosuspend

Make stm32 uart driver use pm_runtime_autosuspend.
Set auto delay to 500ms.

Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I2a21f3cd1775fc4d2a3563c2b613a07ba8a8608f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/414007
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/tty/serial/stm32-usart.c | 153 +++++++++++++++++++++++++------
 drivers/tty/serial/stm32-usart.h |   1 +
 2 files changed, 124 insertions(+), 30 deletions(-)

diff --git a/drivers/tty/serial/stm32-usart.c b/drivers/tty/serial/stm32-usart.c
index e7d30a278e7b..1b3ae55ef78e 100644
--- a/drivers/tty/serial/stm32-usart.c
+++ b/drivers/tty/serial/stm32-usart.c
@@ -709,6 +709,8 @@ static void stm32_usart_rx_dma_complete(void *arg)
 	unsigned int size;
 	unsigned long flags;
 
+	pm_runtime_get(port->dev);
+
 	spin_lock_irqsave(&port->lock, flags);
 	size = stm32_usart_receive_chars(port, false);
 	uart_unlock_and_check_sysrq_irqrestore(port, flags);
@@ -717,6 +719,9 @@ static void stm32_usart_rx_dma_complete(void *arg)
 	if (!stm32port->has_rtor)
 		mod_timer(&stm32port->rx_dma_timer,
 			  jiffies + msecs_to_jiffies(LPUART_RECEIVE_TIMEOUT_MS));
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static void stm32_usart_rx_dma_timer_callback(struct timer_list *timer)
@@ -790,6 +795,10 @@ static void stm32_usart_tx_dma_terminate(struct stm32_port *stm32_port)
 {
 	dmaengine_terminate_async(stm32_port->tx_ch);
 	stm32_port->tx_dma_busy = false;
+
+	/* Balance pm_runtime_get() called when starting the DMA tx */
+	pm_runtime_mark_last_busy(stm32_port->port.dev);
+	pm_runtime_put_autosuspend(stm32_port->port.dev);
 }
 
 static bool stm32_usart_tx_dma_started(struct stm32_port *stm32_port)
@@ -828,10 +837,15 @@ static void stm32_usart_tx_dma_complete(void *arg)
 
 	stm32_usart_tx_dma_terminate(stm32port);
 
+	pm_runtime_get(port->dev);
+
 	/* Let's see if we have pending data to send */
 	spin_lock_irqsave(&port->lock, flags);
 	stm32_usart_transmit_chars(port);
 	spin_unlock_irqrestore(&port->lock, flags);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static void stm32_usart_tx_interrupt_enable(struct uart_port *port)
@@ -944,6 +958,9 @@ static void stm32_usart_transmit_chars_dma(struct uart_port *port)
 	if (!desc)
 		goto fallback_err;
 
+	/* This will be balanced when the DMA transfer will complete */
+	pm_runtime_get(port->dev);
+
 	/*
 	 * Set "tx_dma_busy" flag. This flag will be released when
 	 * dmaengine_terminate_async will be called. This flag helps
@@ -1066,6 +1083,8 @@ static irqreturn_t stm32_usart_interrupt(int irq, void *ptr)
 	unsigned int size;
 	irqreturn_t ret = IRQ_NONE;
 
+	pm_runtime_get(port->dev);
+
 	sr = readl_relaxed(port->membase + ofs->isr);
 	cr1 = readl_relaxed(port->membase + ofs->cr1);
 	cr3 = readl_relaxed(port->membase + ofs->cr3);
@@ -1140,6 +1159,9 @@ static irqreturn_t stm32_usart_interrupt(int irq, void *ptr)
 		ret = IRQ_HANDLED;
 	}
 
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
+
 	return ret;
 }
 
@@ -1148,12 +1170,16 @@ static void stm32_usart_set_mctrl(struct uart_port *port, unsigned int mctrl)
 	struct stm32_port *stm32_port = to_stm32_port(port);
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
 
+	pm_runtime_get(port->dev);
+
 	if ((mctrl & TIOCM_RTS) && (port->status & UPSTAT_AUTORTS))
 		stm32_usart_set_bits(port, ofs->cr3, USART_CR3_RTSE);
 	else
 		stm32_usart_clr_bits(port, ofs->cr3, USART_CR3_RTSE);
 
 	mctrl_gpio_set(stm32_port->gpios, mctrl);
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static unsigned int stm32_usart_get_mctrl(struct uart_port *port)
@@ -1182,12 +1208,17 @@ static void stm32_usart_stop_tx(struct uart_port *port)
 {
 	struct stm32_port *stm32_port = to_stm32_port(port);
 
+	pm_runtime_get(port->dev);
+
 	stm32_usart_tx_interrupt_disable(port);
 
 	/* dma terminate may have been called in case of dma pause failure */
 	stm32_usart_tx_dma_pause(stm32_port);
 
 	stm32_usart_rs485_rts_disable(port);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 /* There are probably characters waiting to be transmitted. */
@@ -1195,12 +1226,18 @@ static void stm32_usart_start_tx(struct uart_port *port)
 {
 	struct circ_buf *xmit = &port->state->xmit;
 
+	pm_runtime_get(port->dev);
+
 	if (uart_circ_empty(xmit) && !port->x_char)
-		return;
+		goto out;
 
 	stm32_usart_rs485_rts_enable(port);
 
 	stm32_usart_transmit_chars(port);
+
+out:
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 /* Flush the transmit buffer. */
@@ -1208,8 +1245,13 @@ static void stm32_usart_flush_buffer(struct uart_port *port)
 {
 	struct stm32_port *stm32_port = to_stm32_port(port);
 
+	pm_runtime_get(port->dev);
+
 	if (stm32_port->tx_ch && stm32_usart_tx_dma_started(stm32_port))
 		stm32_usart_tx_dma_terminate(stm32_port);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 /* Throttle the remote when input buffer is about to overflow. */
@@ -1219,6 +1261,17 @@ static void stm32_usart_throttle(struct uart_port *port)
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
 	unsigned long flags;
 
+	/*
+	 * Keep runtime PM active while the port is throttled. Going to a low
+	 * power mode could lead to loose data stored temporarily in FIFO, as
+	 * shouldn't be pushed to upper layer:
+	 * - DMA buffer can't be flushed right now, so DMA channel is kept
+	 *   active.
+	 * - The low power procedure flushes DMA buffer in tty input buffer
+	 *   before it terminates DMA channel.
+	 */
+	pm_runtime_get(port->dev);
+
 	spin_lock_irqsave(&port->lock, flags);
 
 	/*
@@ -1257,6 +1310,10 @@ static void stm32_usart_unthrottle(struct uart_port *port)
 		stm32_usart_rx_dma_start_or_resume(port);
 
 	spin_unlock_irqrestore(&port->lock, flags);
+
+	/* Port is unthrottled, balance pm_runtime */
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 /* Receive stop */
@@ -1265,12 +1322,17 @@ static void stm32_usart_stop_rx(struct uart_port *port)
 	struct stm32_port *stm32_port = to_stm32_port(port);
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
 
+	pm_runtime_get(port->dev);
+
 	/* Disable DMA request line. */
 	stm32_usart_rx_dma_pause(stm32_port);
 
 	stm32_usart_clr_bits(port, ofs->cr1, stm32_port->cr1_irq);
 	if (stm32_port->cr3_irq)
 		stm32_usart_clr_bits(port, ofs->cr3, stm32_port->cr3_irq);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static void stm32_usart_break_ctl(struct uart_port *port, int break_state)
@@ -1279,6 +1341,8 @@ static void stm32_usart_break_ctl(struct uart_port *port, int break_state)
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
 	unsigned long flags;
 
+	pm_runtime_get(port->dev);
+
 	spin_lock_irqsave(&port->lock, flags);
 
 	if (break_state)
@@ -1287,6 +1351,9 @@ static void stm32_usart_break_ctl(struct uart_port *port, int break_state)
 		stm32_usart_clr_bits(port, ofs->rqr, USART_RQR_SBKRQ);
 
 	spin_unlock_irqrestore(&port->lock, flags);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static int stm32_usart_startup(struct uart_port *port)
@@ -1298,9 +1365,11 @@ static int stm32_usart_startup(struct uart_port *port)
 	u32 val;
 	int ret;
 
+	pm_runtime_get(port->dev);
+
 	ret = request_irq(port->irq, stm32_usart_interrupt, 0, name, port);
 	if (ret)
-		return ret;
+		goto out;
 
 	if (stm32_port->swap) {
 		val = readl_relaxed(port->membase + ofs->cr2);
@@ -1317,7 +1386,7 @@ static int stm32_usart_startup(struct uart_port *port)
 		ret = stm32_usart_rx_dma_start_or_resume(port);
 		if (ret) {
 			free_irq(port->irq, port);
-			return ret;
+			goto out;
 		}
 	}
 
@@ -1325,7 +1394,11 @@ static int stm32_usart_startup(struct uart_port *port)
 	val = stm32_port->cr1_irq | USART_CR1_RE | BIT(cfg->uart_enable_bit);
 	stm32_usart_set_bits(port, ofs->cr1, val);
 
-	return 0;
+out:
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
+
+	return ret;
 }
 
 static void stm32_usart_shutdown(struct uart_port *port)
@@ -1336,6 +1409,8 @@ static void stm32_usart_shutdown(struct uart_port *port)
 	u32 val, isr;
 	int ret;
 
+	pm_runtime_get(port->dev);
+
 	if (stm32_usart_tx_dma_started(stm32_port))
 		stm32_usart_tx_dma_terminate(stm32_port);
 
@@ -1372,6 +1447,9 @@ static void stm32_usart_shutdown(struct uart_port *port)
 
 	stm32_usart_clr_bits(port, ofs->cr1, val);
 
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
+
 	free_irq(port->irq, port);
 }
 
@@ -1390,6 +1468,8 @@ static void stm32_usart_set_termios(struct uart_port *port,
 	unsigned long flags;
 	int brr_fit = 0, ret;
 
+	pm_runtime_get(port->dev);
+
 	if (!stm32_port->hw_flow_control)
 		cflag &= ~CRTSCTS;
 
@@ -1622,6 +1702,9 @@ static void stm32_usart_set_termios(struct uart_port *port,
 		stm32_usart_enable_ms(port);
 	else
 		stm32_usart_disable_ms(port);
+
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
 }
 
 static const char *stm32_usart_type(struct uart_port *port)
@@ -1662,13 +1745,14 @@ static void stm32_usart_pm(struct uart_port *port, unsigned int state,
 
 	switch (state) {
 	case UART_PM_STATE_ON:
-		pm_runtime_get_sync(port->dev);
+		clk_prepare_enable(stm32port->clk);
 		break;
 	case UART_PM_STATE_OFF:
 		spin_lock_irqsave(&port->lock, flags);
 		stm32_usart_clr_bits(port, ofs->cr1, BIT(cfg->uart_enable_bit));
 		spin_unlock_irqrestore(&port->lock, flags);
-		pm_runtime_put_sync(port->dev);
+
+		clk_disable_unprepare(stm32port->clk);
 		break;
 	}
 }
@@ -2104,23 +2188,19 @@ static int stm32_usart_serial_probe(struct platform_device *pdev)
 
 	platform_set_drvdata(pdev, &stm32port->port);
 
-	pm_runtime_get_noresume(&pdev->dev);
-	pm_runtime_set_active(&pdev->dev);
-	pm_runtime_enable(&pdev->dev);
-
 	ret = uart_add_one_port(&stm32_usart_driver, &stm32port->port);
 	if (ret)
-		goto err_port;
+		goto err_rtor;
+
+	pm_runtime_set_active(&pdev->dev);
+	pm_runtime_use_autosuspend(&pdev->dev);
+	pm_runtime_set_autosuspend_delay(&pdev->dev, STM32_USART_AUTOSUSPEND_DELAY_MS);
+	pm_runtime_enable(&pdev->dev);
 
-	pm_runtime_put_sync(&pdev->dev);
+	clk_disable_unprepare(stm32port->clk);
 
 	return 0;
 
-err_port:
-	pm_runtime_disable(&pdev->dev);
-	pm_runtime_set_suspended(&pdev->dev);
-	pm_runtime_put_noidle(&pdev->dev);
-
 err_rtor:
 	if (stm32port->tx_ch)
 		stm32_usart_of_dma_tx_remove(stm32port, pdev);
@@ -2154,7 +2234,8 @@ static int stm32_usart_serial_remove(struct platform_device *pdev)
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
 	u32 cr3;
 
-	pm_runtime_get_sync(&pdev->dev);
+	clk_prepare_enable(stm32_port->clk);
+
 	uart_remove_one_port(&stm32_usart_driver, port);
 
 	pm_runtime_disable(&pdev->dev);
@@ -2224,6 +2305,8 @@ static void stm32_usart_console_write(struct console *co, const char *s,
 	else
 		spin_lock_irqsave(&port->lock, flags);
 
+	pm_runtime_get(port->dev);
+
 	/* Save and disable interrupts, enable the transmitter */
 	old_cr1 = readl_relaxed(port->membase + ofs->cr1);
 	new_cr1 = old_cr1 & ~USART_CR1_IE_MASK;
@@ -2235,6 +2318,9 @@ static void stm32_usart_console_write(struct console *co, const char *s,
 	/* Restore interrupt state */
 	writel_relaxed(old_cr1, port->membase + ofs->cr1);
 
+	pm_runtime_mark_last_busy(port->dev);
+	pm_runtime_put_autosuspend(port->dev);
+
 	if (locked)
 		spin_unlock_irqrestore(&port->lock, flags);
 }
@@ -2353,7 +2439,7 @@ static int __maybe_unused stm32_usart_serial_en_wakeup(struct uart_port *port,
 	unsigned int size = 0;
 	unsigned long flags;
 
-	if (!stm32_port->wakeup_src || !tty_port_initialized(tport))
+	if (!device_can_wakeup(port->dev) || !tty_port_initialized(tport))
 		return 0;
 
 	/*
@@ -2381,12 +2467,24 @@ static int __maybe_unused stm32_usart_serial_en_wakeup(struct uart_port *port,
 			uart_unlock_and_check_sysrq_irqrestore(port, flags);
 			if (size)
 				tty_flip_buffer_push(tport);
+
+			stm32_usart_clr_bits(port, ofs->cr3, USART_CR3_DMAR);
+		}
+
+		if (stm32_port->tx_ch) {
+			if (stm32_usart_tx_dma_started(stm32_port))
+				stm32_usart_tx_dma_terminate(stm32_port);
+			stm32_usart_clr_bits(port, ofs->cr3, USART_CR3_DMAT);
 		}
 
 		/* Poll data from RX FIFO if any */
 		stm32_usart_receive_chars(port, false);
 	} else {
+		if (stm32_port->tx_ch)
+			stm32_usart_set_bits(port, ofs->cr3, USART_CR3_DMAT);
+
 		if (stm32_port->rx_ch) {
+			stm32_usart_set_bits(port, ofs->cr3, USART_CR3_DMAR);
 			ret = stm32_usart_rx_dma_start_or_resume(port);
 			if (ret)
 				return ret;
@@ -2407,10 +2505,9 @@ static int __maybe_unused stm32_usart_serial_suspend(struct device *dev)
 	struct stm32_port *stm32port = to_stm32_port(port);
 	int ret;
 
-	uart_suspend_port(&stm32_usart_driver, port);
 
 	if (device_may_wakeup(dev) || device_wakeup_path(dev)) {
-		ret = stm32_usart_serial_en_wakeup(port, true);
+		ret = pm_runtime_force_suspend(dev);
 		if (ret)
 			return ret;
 
@@ -2418,6 +2515,8 @@ static int __maybe_unused stm32_usart_serial_suspend(struct device *dev)
 			stm32_usart_enable_am(stm32port);
 	}
 
+	uart_suspend_port(&stm32_usart_driver, port);
+
 	/*
 	 * When "no_console_suspend" is enabled, keep the pinctrl default state
 	 * and rely on bootloader stage to restore this state upon resume.
@@ -2446,7 +2545,7 @@ static int __maybe_unused stm32_usart_serial_resume(struct device *dev)
 		if (stm32port->clk_am)
 			stm32_usart_disable_am(stm32port);
 
-		ret = stm32_usart_serial_en_wakeup(port, false);
+		ret = pm_runtime_force_resume(dev);
 		if (ret)
 			return ret;
 	}
@@ -2457,21 +2556,15 @@ static int __maybe_unused stm32_usart_serial_resume(struct device *dev)
 static int __maybe_unused stm32_usart_runtime_suspend(struct device *dev)
 {
 	struct uart_port *port = dev_get_drvdata(dev);
-	struct stm32_port *stm32port = container_of(port,
-			struct stm32_port, port);
-
-	clk_disable_unprepare(stm32port->clk);
 
-	return 0;
+	return stm32_usart_serial_en_wakeup(port, true);
 }
 
 static int __maybe_unused stm32_usart_runtime_resume(struct device *dev)
 {
 	struct uart_port *port = dev_get_drvdata(dev);
-	struct stm32_port *stm32port = container_of(port,
-			struct stm32_port, port);
 
-	return clk_prepare_enable(stm32port->clk);
+	return stm32_usart_serial_en_wakeup(port, false);
 }
 
 static const struct dev_pm_ops stm32_serial_pm_ops = {
diff --git a/drivers/tty/serial/stm32-usart.h b/drivers/tty/serial/stm32-usart.h
index 27ac58dc051f..aff60429fb0a 100644
--- a/drivers/tty/serial/stm32-usart.h
+++ b/drivers/tty/serial/stm32-usart.h
@@ -219,6 +219,7 @@ static const unsigned int STM32_USART_PRESC_VAL[] = {1, 2, 4, 6, 8, 10, 12, 16,
 #define TX_BUF_L RX_BUF_L	 /* dma tx buffer length     */
 
 #define STM32_USART_TIMEOUT_USEC USEC_PER_SEC /* 1s timeout in µs */
+#define STM32_USART_AUTOSUSPEND_DELAY_MS 500
 
 #define LPUART_RECEIVE_TIMEOUT_MS 1000
 
-- 
2.50.1

